creator-node/src/routes/tracks.js

```diff
@@ -550,26 +551,99 @@ module.exports = function (app) {
-      order: [['clock', 'DESC']]
-    })
-
-    if (!blockchainIdFromTrack) {
-      return sendResponse(req, res, errorResponseBadRequest(`No track found for blockchainId ${blockchainId}`))
-    }
-
-    const { multihash } = await models.File.findOne({
+    let fileRecord = await models.File.findOne({
       attributes: ['multihash'],
       where: {
         type: 'copy320',
-        trackBlockchainId: blockchainIdFromTrack
+        trackBlockchainId: blockchainId
       },
       order: [['clock', 'DESC']]
     })
 
-    if (!multihash) {
+    try {
+      // see if there's a fileRecord in redis so we can short circuit all this logic
+      let redisFileRecord = await redisClient.get(`streamFallback:::${blockchainId}`)
+      if (redisFileRecord) {
+        redisFileRecord = JSON.parse(redisFileRecord)
+        if (redisFileRecord && redisFileRecord.multihash) {
+          fileRecord = redisFileRecord
+        }
+      }
+    } catch (e) {
+      req.logger.error(`Error looking for stream fallback in redis`, e)
+    }
+
+    // if track didn't finish the upload process and was never associated, there may not be a trackBlockchainId for the File records,
+    // try to fall back to discovery to fetch the metadata multihash and see if you can deduce the copy320 file
+    if (!fileRecord) {
+      try {
+        let trackRecord = await libs.Track.getTracks(1, 0, [blockchainId])
+        if (!trackRecord || trackRecord.length === 0 || !trackRecord[0].hasOwnProperty('blocknumber')) {
+          // TODO - return a server err response
+          return sendResponse(req, res, errorResponseServerError('Missing or malformatted track fetched from discovery node.'))
+        }
+
+        trackRecord = trackRecord[0]
+
+        // query the files table for the metadata multihash from discovery
+        // no CNodeUserUUID, but because this track is associated with a user and contains that user_id
+        // inside it, it's unique to this user
+        const file = await models.File.findOne({ where: { multihash: trackRecord.metadata_multihash } })
+        if (!file) {
+          return sendResponse(req, res, errorResponseServerError('Missing or malformatted track fetched from discovery node.'))
+        }
+
+        // read the metadata multihash from from the file system
+        let metadataJSON
+
+        const fileBuffer = await readFile(file.storagePath)
+        metadataJSON = JSON.parse(fileBuffer)
+        if (
+          !metadataJSON ||
+          !metadataJSON.track_segments ||
+          !Array.isArray(metadataJSON.track_segments) ||
+          !metadataJSON.track_segments.length
+        ) {
+          return sendResponse(req, res, errorResponseServerError(`Malformatted metadataJSON stored for metadata multihash ${trackRecord.metadata_multihash}.`))
+        }
+
+        // make sure all track segments have the same sourceFile
+        const segments = metadataJSON.track_segments.map(segment => segment.multihash)
```

Comment:
why all these sanity checks? it would seem maybe better to err on the side of it being ok if the segments don't line up? / does this really happen often?

Reply:
it should almost never happen but this is useful bc if we pull the sourcefile off any random multihash file entry, it could map to multiple tracks (only if all uploaded by same cnodeUSerUUID) and we could end up streaming the wrong track. given that this is infrequent it seems valuable to do? like all of this should be < 1 sec only the first time request before its cached

but also can see coutnerarg

Reply:
the only case when this should happen is if the user uploads the same song multiple times and it remains unassociated after multiple failed attempts. we can just treat this as warnings and log out and continue, but the tradeoff is if there was a remix etc, there's an edge case where we could play the wrong song if a CID overlaps